designs/NuGetExtensibility/INuGetProjectServices.md

```diff
@@ -0,0 +1,138 @@
+    ProjectNotReady,
+
+    /// <summary>Package information could not be retrieved because the project is in an invalid state</summary>
+    /// <remarks>If a project has an invalid target framework value, or a package reference has a version value, NuGet may be unable to generate basic project information, such as requested packages.</remarks>
+    ProjectInvalid
+}
+
+/// <summary>Basic information about a package</summary>
+public sealed class NuGetInstalledPackage
+{
+    /// <summary>The package id.</summary>
+    public string Id { get; }
+
+    /// <summary>The package version</summary>
+    /// <remarks>
+    /// If the project uses packages.config, this will be the installed package version.
+    /// If the project uses PackageReference, this would ideally be the resolved package version, but may be the requested package version.
+    /// </remarks>
+    public string Version { get; }
+
+    // To maximize the chance this class can be converted to a C#9 record type without breaking changes, constructor is not public
+    internal NuGetInstalledPackage();
+}
+
+/// <summary>Factory to create types</summary>
+/// <remarks>Trying to be forwards compatible with what C#9 records are going to be</remarks>
+public static class ContractsFactory
+{
+    /// <summary>Create a <see cref="NuGetInstalledPackage"/></summary>
+    /// <param name="id"></param>
+    /// <param name="version"></param>
+    /// <param name="isDirectDependency"></param>
+    /// <returns></returns>
+    public static NuGetInstalledPackage CreateNuGetInstalledPackage(string id, string version, bool isDirectDependency);
+
+    /// <summary>Create a <see cref="GetInstalledPackageResultStatus"/></summary>
+    /// <param name="status"></param>
+    /// <param name="packages"></param>
+    /// <returns></returns>
+    public static GetInstalledPackagesResult CreateGetInstalledPackagesResult(GetInstalledPackageResultStatus status, IReadOnlyCollection<NuGetInstalledPackage> packages);
+}
+```
+
+## Future Work
+
+* Add direct/transitive dependency information to `NuGetInstalledPackage`.
+
+* Other required APIs: (incomplete list)
+  * InstallPackageAsync
+  * UninstallPackageAsync
+
+## Open Questions
+
+* GetInstalledPackagesAsync
+  * Should returned package version be the resolved version, or the requested range, or the lower bound of the requested range? Leaning towards requested range. It means 
+    * Should we return both requested and resolved versions?  Leaning to no: most components & extensions wouldn't care, plus it doesn't make sense in `packages.config` projects. If wanted, a PackageReference (PR) specific service could be added, which allows it to provide PR specific properties.
+  * Should there be options to wait for nomination, rather than returning error?
+  * If `NuGetInstalledPackage.Version` is the resolved version, do we distinguish between project not nominated vs restore not complete?  What about if restore failed?
```

Comment:
If restore failed and the package is not on disk, i think we just need to provide null in the package path potentially. Or something along those lines to match the current IVs* API.